doc/manual/metaprogramming.rst

```diff
@@ -872,3 +872,217 @@ entirely in Julia. You can read their source and see precisely what they
 do — and all they do is construct expression objects to be inserted into
 your program's syntax tree.
 
+Staged functions
+----------------
+
+*Staged functions* play a similar role as macros, but at a later stage
+between parsing and run-time. Staged functions give the capability to
+generate specialized code depending on the types of their arguments.
+While macros work with expressions at parsing-time and cannot access the
+types of their inputs, a staged function gets expanded at a time when
+the types of the arguments are known, but the function is not yet compiled.
+
+Depending on the types of the arguments, a staged function returns a quoted
+expression which then forms the function body of the specialized function.
```

Comment:
I'm not a computer science expert, but I think you want to be careful about the distinction between functions and methods.  I think this would be more correctly stated as "a staged function returns … which forms the *method* body of the specialized *method*.  I may be wrong or splitting hairs not worth splitting.